In [2]:
#importing the Libraies
import numpy as np
import matplotlib.pyplot
import pandas as pd

In [3]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [4]:
%cd /content/drive/MyDrive/Colab Notebooks

/content/drive/MyDrive/Colab Notebooks


In [5]:
dataset = pd.read_csv('50_Startups.csv')

In [6]:
dataset

,R&D Spend,Administration,Marketing Spend,State,Profit
0,165349.20,136897.80,471784.10,New York,192261.83
1,162597.70,151377.59,443898.53,California,191792.06
2,153441.51,101145.55,407934.54,Florida,191050.39
3,144372.41,118671.85,383199.62,New York,182901.99
4,142107.34,91391.77,366168.42,Florida,166187.94
5,131876.90,99814.71,362861.36,New York,156991.12
6,134615.46,147198.87,127716.82,California,156122.51
7,130298.13,145530.06,323876.68,Florida,155752.60
8,120542.52,148718.95,311613.29,New York,152211.77
9,123334.88,108679.17,304981.62,California,149759.96


In [8]:
datasets

,R&D Spend,Administration,Marketing Spend,Profit,State_Florida,State_New York
0,165349.20,136897.80,471784.10,192261.83,False,True
1,162597.70,151377.59,443898.53,191792.06,False,False
2,153441.51,101145.55,407934.54,191050.39,True,False
3,144372.41,118671.85,383199.62,182901.99,False,True
4,142107.34,91391.77,366168.42,166187.94,True,False
5,131876.90,99814.71,362861.36,156991.12,False,True
6,134615.46,147198.87,127716.82,156122.51,False,False
7,130298.13,145530.06,323876.68,155752.60,True,False
8,120542.52,148718.95,311613.29,152211.77,False,True
9,123334.88,108679.17,304981.62,149759.96,False,False


In [7]:
datasets=pd.get_dummies(dataset,drop_first=True)

In [9]:
indep=datasets[['R&D Spend','Administration', 'Marketing Spend','State_Florida', 'State_New York']]
dep=datasets[['Profit']]

In [15]:
dependent = dataset[["Profit"]]

In [16]:
#split into training set and test
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(indep, dep, test_size = 1/3, random_state = 0)

In [17]:
X_train

,R&D Spend,Administration,Marketing Spend,State_Florida,State_New York
45,1000.23,124153.04,1903.93,False,True
48,542.05,51743.15,0.00,False,True
29,65605.48,153032.06,107138.38,False,True
15,114523.61,122616.84,261776.23,False,True
30,61994.48,115641.28,91131.24,True,False
32,63408.86,129219.61,46085.25,False,False
16,78013.11,121597.55,264346.06,False,False
42,23640.93,96189.63,148001.11,False,False
20,76253.86,113867.30,298664.47,False,False
43,15505.73,127382.30,35534.17,False,True


In [21]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

In [22]:
X_train

array([[-1.41859108,  0.08007161, -1.46280061, -0.42257713,  1.24034735],
       [-1.42804718, -2.75105709, -1.4772005 , -0.42257713,  1.24034735],
       [-0.08524282,  1.20920221, -0.66688679, -0.42257713,  1.24034735],
       [ 0.92434855,  0.02000827,  0.50267699, -0.42257713,  1.24034735],
       [-0.15976805, -0.25272669, -0.7879527 ,  2.36643191, -0.80622577],
       [-0.13057752,  0.27816764, -1.12864651, -0.42257713, -0.80622577],
       [ 0.17083067, -0.01984459,  0.52211324, -0.42257713, -0.80622577],
       [-0.95132351, -1.01325988, -0.357832  , -0.42257713, -0.80622577],
       [ 0.13452258, -0.3220869 ,  0.78167175, -0.42257713, -0.80622577],
       [-1.11922093,  0.20633131, -1.2084469 , -0.42257713,  1.24034735],
       [ 1.04856916,  1.04056545,  0.87960683, -0.42257713,  1.24034735],
       [ 0.45934048,  0.52352964,  0.433766  , -0.42257713, -0.80622577],
       [-0.1046588 ,  0.68219546, -0.43375555, -0.42257713, -0.80622577],
       [ 1.2824925 , -0.87152415,  1.2

In [24]:
from sklearn.svm import SVR
regressor = SVR(kernel="rbf", C = 1000 )
regressor.fit(X_train,y_train)

/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:1408: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


SVR(C=1000)

In [26]:
print(regressor)

SVR(C=1000)


In [29]:
SVR (C = 1000, cache_size=200, coef0=0.0, degree=3, epsilon=0.1,
     gamma='auto_deprecated', kernel='rbf', max_iter=-1, shrinking=True,
     tol = 0.001, verbose=False)

SVR(C=1000, gamma='auto_deprecated')

In [30]:
regressor.intercept_

array([108085.06854154])

# It shows the intercept (constant / starting value) of a trained Linear Regression model.

Simple Meaning

In linear regression, equation is:

𝑦
=
𝑚
𝑥
+
𝑐
y=mx+c

Where:

m = slope / coefficient
c = intercept

✅ regressor.intercept_ = c value

In [31]:
regressor.n_support_

array([33], dtype=int32)

# This is NOT for Linear Regression usually.
# It is mostly used with Support Vector Machine (SVM / SVR) models.

What it shows

n_support_ tells you:

How many support vectors are used by the trained SVM model

Your output
array([33], dtype=int32)
Meaning:

**Your model is using 33 support vectors**

----------

What is a support vector?

In SVM / SVR:

model draws a best boundary / best fitting margin
only some important data points decide that boundary
those important points are called support vectors

 These are the key training points that influence the model most.

In [32]:
regressor.support_

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32],
      dtype=int32)

regressor.support_ = **It gives the index positions of the support vectors in your training dataset.**

In [33]:
y_pred = regressor.predict(X_test)

In [34]:
from sklearn.metrics import r2_score
r_score=r2_score(y_test,y_pred)

In [37]:
r_score

-0.08004178386944938

In [38]:
import pickle
filename = "finalized_model_SVM_linear.sav"

**pickle** is used to save a trained model into a file and load it later without training it again.


This saves time, because the model can be reused directly for prediction.

In [39]:
pickle.dump(regressor,open(filename,'wb'))

In [44]:
preinput = sc.transform([[1300,12000,4000,0,1]])

/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(


In [45]:
preinput

array([[-1.41240431, -4.30496048, -1.44694752, -0.42257713,  1.24034735]])

In [48]:
loaded_model=pickle.load(open("finalized_model_SVM_linear.sav", 'rb'))
result=loaded_model.predict(preinput)

In [49]:
result

array([107183.38309654])

***Deployment  - preinput***

In [52]:

dataset = pd.read_csv('50_Startups.csv')
datasets=pd.get_dummies(dataset,drop_first=True)
indep=datasets[['R&D Spend','Administration', 'Marketing Spend','State_Florida', 'State_New York']]
dep=datasets[['Profit']]

#split into training set and test
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(indep, dep, test_size = 1/3, random_state = 0)

from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

preinput = sc.transform([[1300,12000,4000,0,1]])


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(


In [53]:
preinput

array([[-1.41240431, -4.30496048, -1.44694752, -0.42257713,  1.24034735]])

In [54]:
loaded_model=pickle.load(open("finalized_model_SVM_linear.sav", 'rb'))
result=loaded_model.predict(preinput)

In [55]:
result

array([107183.38309654])

(1) pickle.load(open("finalized_model_SVM_linear.sav", 'rb')) loads the saved trained SVM model from the .sav file.

(2) This means the model is already trained and ready to use for prediction.

(3) loaded_model.predict(preinput) sends the new input data to the loaded model.

(4) The model processes the input values and predicts the output based on what it learned during training.

(5) result = array([107183.38309654]) means the predicted value for the given input is 107183.38.


===========

opens the saved file

reads the trained SVM model

loads it back into Python


SVR (Support Vector Regression) is a machine learning algorithm used to predict continuous values like salary, price, marks, temperature, etc.

Simple meaning:
SVM = used for classification (Yes/No, Cat/Dog)
SVR = used for regression (predicting a number)
Example:

If you give:

experience = 5 years
skill = 8

SVR can predict:

salary = ₹50,000
One-line:

SVR is used when the output is a number, not a category.

**If output is class/category → use SVM**
If output is continuous number → use SVR